src/Pate/Proof.hs

```diff
@@ -83,286 +114,343 @@ data EquivTripleBody sym arch where
+  | VerificationSkipped
+  | VerificationSuccess
+  | VerificationFail ce
+  deriving (Functor, Foldable, Traversable)
+
+
+-- | An abstract proof object, representing the overall structure of an equivalence proof.
+data ProofApp prf (node :: ProofNodeType -> *) (tp :: ProofNodeType) where
+  -- | Proof that the top-level triple is satisfied, according to all possible
+  -- exits from the given block slice
+  ProofBlockSlice ::
+    { prfBlockSliceTriple :: node ProofTripleType
+    , prfBlockSliceCalls :: [(PT.PatchPair (ProofBlock prf), node ProofFunctionCallType)]
+    , prfBlockSliceReturn :: Maybe (node ProofTripleType)
+    , prfBlockSliceUnknownExit :: Maybe (node ProofTripleType)
+    , prfBlockSliceTrans :: BlockSliceTransition prf
+    } -> ProofApp prf node ProofBlockSliceType
+
+  ProofFunctionCall ::
+    { prfFunctionCallPre :: node ProofTripleType
+    , prfFunctionCallBody :: node ProofBlockSliceType
+    , prfFunctionCallContinue :: Maybe (node ProofBlockSliceType)
+    } -> ProofApp prf node ProofFunctionCallType
+
+  ProofTriple ::
+    { prfTripleBlocks :: PT.PatchPair (ProofBlock prf)
+    , prfTriplePreDomain :: node ProofDomainType
+    , prfTriplePostDomain :: node ProofDomainType
+    , prfTripleStatus :: node ProofStatusType
+    
+    } -> ProofApp prf node ProofTripleType
+
+  ProofStatus ::
+    { prfStatus :: VerificationStatus (ProofCounterExample prf) }
+    -> ProofApp prf node ProofStatusType
+
+  ProofDomain ::
+    { prfDomainRegisters :: MM.RegState (ProofRegister prf) (Const (ProofPredicate prf))
+    , prfDomainStackMemory :: ProofMemoryDomain prf
+    , prfDomainGlobalMemory :: ProofMemoryDomain prf
+    , prfDomainContext :: ProofContext prf
+    }  -> ProofApp prf node ProofDomainType
+
+
+data ProofMemoryDomain prf where
+  ProofMemoryDomain ::
+    { prfMemoryDomain :: MapF.MapF (ProofMemCell prf) (Const (ProofPredicate prf))
+    , prfMemoryDomainPolarity :: ProofPredicate prf
+    }  -> ProofMemoryDomain prf
+
+traverseProofApp ::
+  Applicative m =>
+  (forall tp. node tp -> m (node' tp)) ->
+  ProofApp prf node utp ->
+  m ((ProofApp prf node') utp)
+traverseProofApp f = \case
+  ProofBlockSlice a1 a2 a3 a4 a5 -> ProofBlockSlice
+    <$> f a1
+    <*> traverse (\(blks,v) -> (,) <$> pure blks <*> f v) a2
+    <*> traverse f a3
+    <*> traverse f a4
+    <*> pure a5
+  ProofFunctionCall a1 a2 a3 -> ProofFunctionCall
+    <$> f a1
+    <*> f a2
+    <*> traverse f a3
+  ProofTriple a1 a2 a3 a4 -> ProofTriple
+    <$> pure a1
+    <*> f a2
+    <*> f a3
+    <*> f a4
+    
+  ProofStatus a1 -> ProofStatus
+    <$> pure a1
+  ProofDomain a1 a2 a3 a4 -> ProofDomain
+    <$> pure a1
+    <*> pure a2
+    <*> pure a3
+    <*> pure a4
+
+mapProofApp ::
+  (forall tp. node tp -> node' tp) ->
+  ProofApp prf node utp ->
+  ProofApp prf node' utp
+mapProofApp f app = runIdentity $ traverseProofApp (\app' -> Identity $ f app') app
+
+-- | A bundle of functions for converting the leafs of a proof graph
```

Comment:
Is that converting leaves from `prf` to `prf'`?

Reply:
Giving some indication as to what each of these pieces is for would be really useful